server/model.go

```diff
@@ -129,123 +130,137 @@ func detectContentType(r io.Reader) (string, error) {
-		}
-
-		return false
-	})
+	var walk func(nodes []parse.Node)
+	walk = func(nodes []parse.Node) {
+		for _, node := range nodes {
+			if found {
+				return
+			}
 
-	if tmpl == nil {
-		return nil, false
-	}
+			switch n := node.(type) {
+			case *parse.IfNode:
+				slog.Debug("TextAfterToolCalls: processing IfNode", "node", n.String())
+				if nodeContainsToolCalls(n) {
+					slog.Debug("TextAfterToolCalls: found ToolCalls in IfNode")
+					// Collect immediate TextNode(s) at start of IfNode's list
+					var sb strings.Builder
+					for _, innerNode := range n.List.Nodes {
+						if tn, ok := innerNode.(*parse.TextNode); ok {
+							slog.Debug("TextAfterToolCalls: found TextNode", "text", string(tn.Text))
+							sb.Write(tn.Text)
+						} else {
+							// Stop at first non-text node
+							break
+						}
+					}
+					result = sb.String()
+					found = true
+					slog.Debug("TextAfterToolCalls: found result", "result", result)
+					return
+				}
+				// Recurse into child nodes
+				walk(n.List.Nodes)
+				if n.ElseList != nil {
+					walk(n.ElseList.Nodes)
+				}
+			case *parse.ListNode:
+				slog.Debug("TextAfterToolCalls: processing ListNode")
+				walk(n.Nodes)
+			case *parse.RangeNode:
+				slog.Debug("TextAfterToolCalls: processing RangeNode")
+				walk(n.List.Nodes)
+				if n.ElseList != nil {
+					walk(n.ElseList.Nodes)
+				}
+			case *parse.WithNode:
+				slog.Debug("TextAfterToolCalls: processing WithNode")
+				walk(n.List.Nodes)
+				if n.ElseList != nil {
+					walk(n.ElseList.Nodes)
+				}
+			default:
+				// Continue to next node
+				continue
+			}
 
-	var b bytes.Buffer
-	if err := tmpl.Execute(&b, map[string][]api.ToolCall{
-		"ToolCalls": {
-			{
-				Function: api.ToolCallFunction{
-					Name: "@@name@@",
-					Arguments: api.ToolCallFunctionArguments{
-						"@@argument@@": 1,
-					},
-				},
-			},
-		},
-	}); err != nil {
-		return nil, false
+			if found {
+				return
+			}
+		}
 	}
 
-	templateObjects := parseObjects(b.String())
-	if len(templateObjects) == 0 {
-		return nil, false
-	}
+	walk(tmpl.Tree.Root.Nodes)
+	slog.Debug("TextAfterToolCalls: finished walking tree", "result", result, "found", found)
+	return result, found
+}
 
-	// find the keys that correspond to the name and arguments fields
-	var name, arguments string
-	for k, v := range templateObjects[0] {
-		switch v.(type) {
-		case string:
-			name = k
-		case map[string]any:
-			arguments = k
+// Helper to detect if a node's condition includes ".ToolCalls"
+func nodeContainsToolCalls(n *parse.IfNode) bool {
+	for _, cmd := range n.Pipe.Cmds {
+		for _, arg := range cmd.Args {
+			if field, ok := arg.(*parse.FieldNode); ok {
+				if slices.Contains(field.Ident, "ToolCalls") {
+					return true
+				}
+			}
 		}
 	}
+	return false
+}
 
-	if name == "" || arguments == "" {
-		return nil, false
-	}
-
-	responseObjects := parseObjects(s)
-	if len(responseObjects) == 0 {
-		return nil, false
+func ToolToken(found string) (string, bool) {
```

Comment:
Does this need to be Exposed?

Reply:
I tried doing this with the template portion under `model.go` but it made testing more difficult with having to load the template correctly. Going to take another stab at it but not sure if it'll work